src/MemoryTracing/include/MemoryTracing/MemoryInfoListener.h

```diff
@@ -13,9 +17,24 @@ namespace orbit_memory_tracing {
 class MemoryInfoListener {
  public:
   virtual ~MemoryInfoListener() = default;
-  virtual void OnSystemMemoryUsage(orbit_grpc_protos::SystemMemoryUsage memory_info) = 0;
-  virtual void OnProcessMemoryUsage(orbit_grpc_protos::ProcessMemoryUsage memory_info) = 0;
-  virtual void OnCGroupMemoryUsage(orbit_grpc_protos::CGroupMemoryUsage memory_info) = 0;
+
+  void OnSystemMemoryUsage(orbit_grpc_protos::SystemMemoryUsage system_memory_usage);
+  void OnCGroupMemoryUsage(orbit_grpc_protos::CGroupMemoryUsage cgroup_memory_usage);
+  void OnProcessMemoryUsage(orbit_grpc_protos::ProcessMemoryUsage process_memory_usage);
+
+ protected:
+  // This method compute the arithmetic mean of input timestamps.
+  [[nodiscard]] static uint64_t GetSynchronizedSamplingTimestamp(
+      const std::vector<uint64_t>& sampling_timestamps);
+
+  absl::flat_hash_map<uint64_t, orbit_grpc_protos::MemoryUsageEvent>
+      in_progress_memory_usage_events_ GUARDED_BY(in_progress_memory_usage_events_mutex_);
+  absl::Mutex in_progress_memory_usage_events_mutex_;
+  uint64_t sampling_start_timestamp_ns_;
+  uint64_t sampling_period_ns_;
+
+ private:
+  virtual void ProcessMemoryUsageEventIfReady(uint64_t sampling_window_id) = 0;
```

Comment:
I am quite skeptical about the design of this class. The subclasses should just have to implement a method to handle the complete `ProducerCaptureEvent` (e.g., `OnMemoryUsageEvent`) containing the `MemoryUsageEvent`. `ProcessMemoryUsageEventIfReady` should be private here and implemented, and call `OnMemoryUsageEvent` when the event is ready. 

Instead, right now the subclasses now have to implement a lot of very specific logic based on `protected` fields and methods, which should be `private` instead. Consider how much logic is in `BufferMemoryInfoListener`, which should instead just be a very simple class, as it's just a fake for tests.

Reply:
Thanks for the suggestion! Changed accordingly.